### ISSUE TITLE
Nodes not able to discover themselves using aws discovery in 1.2.0 

### ISSUE THREAD

**Issue by petreboy14** [crate/crate#5428]
**CrateDB version**:
1.2.0

**JVM version**: 1.8.0_121-b13
```
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-b13)
OpenJDK 64-Bit Server VM (build 25.121-b13, mixed mode)
```
**OS version / environment description**: 
Amazon Linux ami-920798f2 (The official Crate 1.2.0 AMI). We are running 3 m4.medium servers to test the upgrade from 1.1.1 to 1.2.0. These are all located in the same availability zone, in the same region. They share the same security group, IAM user, and a matching tag. 

The only modifications to crate.yml are the following (for both 1.1.1 and 1.2.0):
```
discovery.type: ec2
discovery.ec2.groups: dev-crate-instance-group
discovery.ec2.tag.crate-env: dev
discovery.zen.minimum_master_nodes: 2
gateway:
  recover_after_nodes: 3
  recover_after_time: 5m
  expected_nodes: 3
```

**Problem description**:

When the nodes are started in version 1.2.0 the cannot complete the discovery process. We see the following error show up in the startup logs:
```
[ec2-user@ip-172-30-1-180 ~]$ sudo tail -f /var/log/crate/crate.log
[2017-04-26T20:21:50,178][INFO ][i.c.rest                 ] [Montagne Durbonas] Elasticsearch HTTP REST API not enabled
[2017-04-26T20:21:50,195][INFO ][o.e.b.BootstrapProxy$1   ] [Montagne Durbonas] initialized
[2017-04-26T20:21:50,195][INFO ][o.e.n.Node               ] [Montagne Durbonas] starting ...
[2017-04-26T20:21:50,247][INFO ][psql                     ] [Montagne Durbonas] publish_address {127.0.0.1:5432}, bound_addresses {127.0.0.1:5432}
[2017-04-26T20:21:50,247][INFO ][i.c.b.BlobService        ] [Montagne Durbonas] BlobService.doStart() io.crate.blob.BlobService@1a8e44fe
[2017-04-26T20:21:50,272][INFO ][o.e.h.HttpServer         ] [Montagne Durbonas] publish_address {127.0.0.1:4200}, bound_addresses {[::1]:4200}, {127.0.0.1:4200}
[2017-04-26T20:21:50,303][INFO ][o.e.t.TransportService   ] [Montagne Durbonas] publish_address {127.0.0.1:4300}, bound_addresses {[::1]:4300}, {127.0.0.1:4300}
[2017-04-26T20:22:20,336][WARN ][o.e.n.Node               ] [Montagne Durbonas] timed out while waiting for initial discovery state - timeout: 30s
[2017-04-26T20:22:20,336][INFO ][o.e.n.Node               ] [Montagne Durbonas] started
```
This occurs on all nodes. We've double checked that they can reach each other and they can over both 4200 and 4300.

If we attempt to curl the index page of one of the nodes we get a response like this:
```
{
  "ok" : false,
  "status" : 503,
  "name" : "Elm",
  "cluster_name" : "crate",
  "version" : {
    "number" : "1.2.0",
    "build_hash" : "af006fa24762e47da523e09181258a7a3cda5849",
    "build_timestamp" : "2017-04-24T11:58:22Z",
    "build_snapshot" : false,
    "es_version" : "5.0.2",
    "lucene_version" : "6.2.1"
  }
}
```
After we were unable to upgrade an existing cluster we tried to roll out a brand new cluster with the AMI mentioned above and experienced the exact same issue. Rolling back to 1.1.1 and the cluster came up fine including discovering themselves. 

**Steps to reproduce**:
1. Launch 2 or more ami-920798f2 in the same region, subnet.
2. Ensure that they have the same security group, one shared tag and that the security group allows traffic on 4300 between themselves. Also ensure they have a shared IAM user with ec2 describe instances. 
3. Modify the /etc/crate/crate.yml with the aforementioned configuration.
4. The nodes should not be able to discover themselves. 

Thank you for the help and please let me know if I need to add more details anywhere.

**Comment by chaudum** [comment]
hi @petreboy14 
CrateDB binds default to the loopback interface since 1.2 (see changelog https://crate.io/docs/reference/en/1.2.0/release_notes/1.2.0.html?highlight=loopback)
you need to bind and publish the site-local address of the host by setting

```
network.host: _site_
```

from the `crate.yml`:
```
############################## Network And HTTP ###############################

# Crate, by default, binds itself to the loopback addresses on the system, and listens
# on port [4200-4300] for HTTP traffic and on port [4300-4400] for node-to-node
# communication. (the range means that if the port is busy, it will automatically
# try the next port).
# Apart from IPv4 and IPv6 addresses there are some special values that can be used:
# _local_ Any loopback addresses on the system, for example 127.0.0.1.
# _site_ Any site-local addresses on the system, for example 192.168.0.1.
# _global_ Any globally-scoped addresses on the system, for example 8.8.8.8.
# _[networkInterface]_  Addresses of a network interface, for example _en0_.
```

**Comment by chaudum** [comment]
However, we will improve the documentation about that.

**Comment by petreboy14** [comment]
Hi @chaudum we will give that a try. Thank you!

**Comment by petreboy14** [comment]
It seems like the problem still remains. 
Here's the startup log:
```
[ec2-user@ip-172-30-1-204 ~]$ tail -f /var/log/crate/crate.log
[2017-04-28T06:48:28,412][INFO ][i.c.rest                 ] [Hochstadel] Elasticsearch HTTP REST API not enabled
[2017-04-28T06:48:28,447][INFO ][o.e.b.BootstrapProxy$1   ] [Hochstadel] initialized
[2017-04-28T06:48:28,448][INFO ][o.e.n.Node               ] [Hochstadel] starting ...
[2017-04-28T06:48:28,560][INFO ][psql                     ] [Hochstadel] publish_address {172.30.1.204:5432}, bound_addresses {172.30.1.204:5432}
[2017-04-28T06:48:28,560][INFO ][i.c.b.BlobService        ] [Hochstadel] BlobService.doStart() io.crate.blob.BlobService@1f293cb7
[2017-04-28T06:48:28,604][INFO ][o.e.h.HttpServer         ] [Hochstadel] publish_address {172.30.1.204:4200}, bound_addresses {172.30.1.204:4200}
[2017-04-28T06:48:28,715][INFO ][o.e.t.TransportService   ] [Hochstadel] publish_address {172.30.1.204:4300}, bound_addresses {172.30.1.204:4300}
[2017-04-28T06:48:28,719][INFO ][o.e.b.BootstrapCheck     ] [Hochstadel] bound or publishing to a non-loopback or non-link-local address, enforcing bootstrap checks
[2017-04-28T06:48:58,739][WARN ][o.e.n.Node               ] [Hochstadel] timed out while waiting for initial discovery state - timeout: 30s
[2017-04-28T06:48:58,739][INFO ][o.e.n.Node               ] [Hochstadel] started
```
Here is the changed configuration file (only the portions that are non-default):
```
discovery.type: ec2
discovery.ec2.groups: dev-crate-instance-group
discovery.ec2.tag.crate-env: dev
discovery.zen.minimum_master_nodes: 2
gateway:
  recover_after_nodes: 3
  recover_after_time: 5m
  expected_nodes: 3
network.host: _site_
```
And then just to show the security groups are open here is the result of curling both 4200 and 4300 from one node to the other:
```
[ec2-user@ip-172-30-1-204 ~]$ curl 172.30.1.45:4200
{
  "ok" : false,
  "status" : 503,
  "name" : "Monte Servin",
  "cluster_name" : "crate",
  "version" : {
    "number" : "1.2.0",
    "build_hash" : "af006fa24762e47da523e09181258a7a3cda5849",
    "build_timestamp" : "2017-04-24T11:58:22Z",
    "build_snapshot" : false,
    "es_version" : "5.0.2",
    "lucene_version" : "6.2.1"
  }
}
[ec2-user@ip-172-30-1-204 ~]$ curl 172.30.1.45:4300
This is not a HTTP port
```

**Comment by chaudum** [comment]
hm, can you increase the logging for discovery by setting

```properties
#  Discovery
#  Crate will discover the other nodes within its own cluster.
#  If you want to log the discovery process, set the following:
logger.discovery.name: discovery
logger.discovery.level: TRACE
```

in the `log4j2.properties`.

**Comment by petreboy14** [comment]
I added that to the log4j2.properties and no change in logs from before. Also wanted to mention these instances are in us-west-2. I'm not sure if that makes a difference

**Comment by petreboy14** [comment]
As an update I got further:
I needed to set: `cloud.aws.ec2.endpoint: https://ec2.us-west-2.amazonaws.com` to be able to start the discovery process. However, now there seem to be issues with the handshake working

```
[2017-04-28T21:46:18,934][WARN ][o.e.d.z.p.u.UnicastZenPing] [Wistätthorn] [12] failed send ping to {#zen_unicast_115_#cloud-i-07de4eaa61859a19d-0#}{tTs4ru9sQ4CAzxn_VhTD6g}{172.30.1.196}{172.30.1.196:4300}
java.lang.IllegalStateException: handshake failed with {#zen_unicast_115_#cloud-i-07de4eaa61859a19d-0#}{tTs4ru9sQ4CAzxn_VhTD6g}{172.30.1.196}{172.30.1.196:4300}
	at org.elasticsearch.transport.TransportService.handshake(TransportService.java:364) ~[crate-app-1.2.0.jar:1.2.0]
	at org.elasticsearch.transport.TransportService.connectToNodeLightAndHandshake(TransportService.java:339) ~[crate-app-1.2.0.jar:1.2.0]
	at org.elasticsearch.transport.TransportService.connectToNodeLightAndHandshake(TransportService.java:313) ~[crate-app-1.2.0.jar:1.2.0]
	at org.elasticsearch.discovery.zen.ping.unicast.UnicastZenPing$2.run(UnicastZenPing.java:419) [crate-app-1.2.0.jar:1.2.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:458) [crate-app-1.2.0.jar:1.2.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
Caused by: org.elasticsearch.transport.NodeDisconnectedException: [][172.30.1.196:4300][internal:transport/handshake] disconnected
```

**Comment by petreboy14** [comment]
Ok we can disregard the last comment. The discovery was conflicting with other tags in the system. Making the cloud.aws.ec2.endpoint change seemed to do the trick! Thanks for the help.

**Comment by chaudum** [comment]
hi @petreboy14 
I've been looking at the relevant code and discovered that the automatic detection of the region has  accidentally been removed with the 1.2 update :(
we'll provide a fix for that in the next 1.2.x release

**Comment by petreboy14** [comment]
@chaudum great news! Looking forward.
